Add `raw-file-link` feature

idasbiste opened the pull request:
This PR closes #1595 
It adds a "View raw" button next to the usual "View file" one.

Without rich diff:
<img width="399" alt="Screenshot 2019-04-13 at 01 22 20" src="https://user-images.githubusercontent.com/5243867/56072037-e2981b00-5d8a-11e9-956f-07acc1479396.png">

With rich diff:
<img width="501" alt="Screenshot 2019-04-13 at 01 22 14" src="https://user-images.githubusercontent.com/5243867/56072034-dd3ad080-5d8a-11e9-9317-20ce2fd328cf.png">

bfred-it commented on source/content.ts:
```diff
@@ -93,6 +93,7 @@ import './features/collapsible-content-button';
 import './features/hide-watch-and-fork-count';
 import './features/resolve-conflicts';
 import './features/follow-file-renames';
+import './features/view-raw-file';
```
The feature should probably be named `raw-file-link`

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+import select from 'select-dom';
+import features from '../libs/features';
+
+const createRawUrl = (href: string) => href.replace(/blob\//, 'raw/');
```
This fails on user/repo containing blob. Use something like:

```js 
url = link.pathname.split('/')
url[3] = 'raw'; // Replaces 'blob'
return url.join('/');

idasbiste replied:
🤦‍♂️ Of course... I was worried about the path or the filename and forgot about the user/repo!

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+		'.file-header .file-actions > div.BtnGroup'
+	);
+
+	for (const group of fileActionsGroups) {
```
You can inline the select in here, on one line. 

Instead of selecting the group, select the link you’ll need on the next line with `[href*='/blob/']` and then use `link.after()` to append our link

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+		const rawUrl = createRawUrl(group.lastElementChild.getAttribute('href'));
+		group.append(
+			<a href={rawUrl} class="btn btn-sm BtnGroup-item" rel="nofollow">
+				View raw
```
GitHub just calls this button Raw, not View raw

idasbiste replied:
You're correct 👍 
I've called it "View raw" because of the discussion on the issue. Happy to change it though for the sake of consistency.

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+	for (const group of fileActionsGroups) {
+		const rawUrl = createRawUrl(group.lastElementChild.getAttribute('href'));
+		group.append(
+			<a href={rawUrl} class="btn btn-sm BtnGroup-item" rel="nofollow">
```
We don’t need nofollow

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+	id: 'view-raw-file',
+	include: [
+		features.isCommit,
+		features.isCommitList,
```
Commit list doesn’t have files

bfred-it commented on source/features/view-raw-file.tsx:
```diff
@@ -0,0 +1,31 @@
+	include: [
+		features.isCommit,
+		features.isCommitList,
+		features.isPRFiles
```
This page may load new files asynchronously. Look at some other features on this page to see how to watch the dom and update it

idasbiste replied:
Good point, thanks! I'll have a look.

idasbiste commented:
@bfred-it I think I've addressed all your comments! Please have a look.
To deal with the asynchronous files load and the respective DOM updates, I've used the MutationObserver. I had to recursively look for elements with a specific class and observe it because the files are added in "cascade" mode inside the second `<div class="js-diff-progressive-container">`. If the PR is large enough you can get 3 or 4 levels of this div. Something like:
```
<div class="js-diff-progressive-container">
  ...
  <div class="js-diff-progressive-container">
    ...
    <div class="js-diff-progressive-container">
     ...
    </div>
  </div>
</div>
```

sindresorhus commented:
@idasbiste Can you rebase from master (we enabled TS strict mode) and fix any issues?

sindresorhus commented:
Seems to work fine for me otherwise.

bfred-it commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,54 @@
+	addRawBtn(select('.js-diff-progressive-container:first-child'));
+	const lastProgressiveContainer = select('.js-diff-progressive-container:last-child');
+	if (lastProgressiveContainer) {
+		observer.observe(lastProgressiveContainer, {childList: true});
```
Maybe we should turn this into a utility function, like `on-new-pr-files.ts`

bfred-it commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,54 @@
+	if (node) {
+		for (const link of select.all('.file-header .file-actions > div.BtnGroup [href*=\'/blob/\']', node)) {
+			const href = link.getAttribute('href');
+			if (href) {
```
We already select only links with `href`, so it always there

idasbiste replied:
It is but after enabling TS strict mode, `getAttribute` returns string or null. This is solved by getting the pathname as mentioned above.

bfred-it replied:
Like in the other comment, the null TS issue can be shut up with `!`, so

```ts
const href = link.getAttribute('href')!;
```

And `href` will always be `string` instead of `string | null`

bfred-it commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,54 @@
+			const href = link.getAttribute('href');
+			if (href) {
+				link.after(
+					<a href={createRawUrl(href)} className="btn btn-sm BtnGroup-item">
```
You can use `link.pathname` instead of `(href)` here, skipping the variable altogether

bfred-it commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,54 @@
+});
+
+function addRawBtn(node: Element | null) {
+	if (node) {
```
As in the other PR, node is always Element now

idasbiste replied:
Unfortunately, the `select` function returns an `HTMLElement` or null. By passing it as argument of this function when calling it in the `init` method, TS strict mode "forces" me to make this check.

bfred-it replied:
You can use `select(selector)!`, the `!` will become your best friend :D 

But https://github.com/sindresorhus/refined-github/commit/a6a6e15d1d0300789b4f548dffb7d5777ad10bc8 avoids this problem too.

idasbiste replied:
I didn't know about the `!`. Thanks!

idasbiste commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,49 @@
+function addRawBtn(node: Element) {
+	for (const link of select.all('.file-header .file-actions > div.BtnGroup [href*=\'/blob/\']', node)) {
+		link.after(
+			<a href={createRawUrl(link.getAttribute('href'))} class="btn btn-sm BtnGroup-item">
```
I wasn't able to get the `pathname` property from a raw Element. I need to cast it to an `HTMLAnchorElement`. Also, I didn't face any problem using the href because the path is always relative.
Does this look ok to you `createRawUrl((link as HTMLAnchorElement).pathname)`?

bfred-it replied:
You can use `select.all<HTMLAnchorElement>` and `link` will automatically be one

idasbiste commented on source/features/raw-view-link.tsx:
```diff
@@ -0,0 +1,49 @@
+
+function init() {
+	addRawBtn(select('.js-diff-progressive-container:first-child'));
+	observer.observe(select('.js-diff-progressive-container:last-child'), childList);
```
It is the last child. Please see my comment above https://github.com/sindresorhus/refined-github/pull/1924#issuecomment-482838249

bfred-it replied:
> It is the last child

In the HTML you posted I see two `:last-child`, and I think `querySelector` will match the first one (level 2), not the one you expect (level 3)

See https://github.com/sindresorhus/refined-github/pull/1932/commits/a6a6e15d1d0300789b4f548dffb7d5777ad10bc8 for an alternative implementation that avoids this altogether

Also do you know a PR that has more than one level? Because in the biggest PR I can think of (https://github.com/sindresorhus/refined-github/pull/1783) I'm seeing only one deferred level, _not_ nested:


```html
<div class="js-diff-progressive-container">already here</div>
<div class="js-diff-progressive-container">(content deferred)</div>
```

idasbiste replied:
I like your approach on the other PR 👍 
In the HTML I've posted I was specifically talking about the container that gets updated -- which is the `:last-child` in the beginning (the one with deferred content). I.e. the HTML will, in the end, have a format of:
```
<div class="js-diff-progressive-container">already here</div>
<div class="js-diff-progressive-container">
  (nested containers)
</div>
```

bfred-it commented:
Please run this to delete the tags from your fork and locally, we keep pulling them from your fork and pushing them to this repo (we no longer use tags)

```sh
git tag -d $(git tag | grep -E '.')
git ls-remote --tags --refs origin | cut -f2 | xargs git push origin --delete
```

bfred-it commented:
I want to merge this but I'm seeing a couple of unrelated RGH buts on this, I want to use this PR as testing ground. I'll merge soon

notlmn commented:
> I want to merge this but I'm seeing a couple of unrelated RGH bugs on this (`show-names` and `co-authors`), I want to use this PR as testing ground. I'll merge soon

`show-names` is broken as the author of this PR has used an email to make commits that do not match with a GitHub account, in that case GH would display only the name used to create that commit. Querying the v4 API with this name (and especially which contain spaces) would fail, hence the broken feature.

#### Non-referencing commit
![image](https://user-images.githubusercontent.com/37769974/56505457-5f6c8700-6539-11e9-813b-46941f7d7d48.png)

#### GraphQL exception
![image](https://user-images.githubusercontent.com/37769974/56505643-c7bb6880-6539-11e9-87c4-e5636d2c769d.png)

----------------------

Didn't want to go off topic but an it is an interesting bug that sometimes bites the GH team too, `add-co-authored-by` is probably broken for the same reason as above (didn't test that though).

bfred-it commented:
Thanks Eduardo! 🎯
